File: js/lib/strings/ReverseStringExercises.js
class ReverseStringExercises {

    /**
     * reverses the array of characters with O(1) space
     * @param {Array} s an array of characters
     */
    static reverseString(s) {
        let li = s.length - 1;

        for (let i = 0; i < Math.floor(s.length / 2); i++) {
            let tmp = s[i];
            s[i] = s[li - i];
            s[li - i] = tmp;
        } 
    }

    /**
     * reverses the array of characters with O(1) space (recursive version)
     * @param {Array} s an array of characters
     */
    static reverseString2(s) {
        return this.reverseStringRecursion(s, 0, s[0]);
    }

    static reverseStringRecursion(s, idx, tmp) {
        s[idx] = s[(s.length - 1) - idx];
        s[(s.length - 1) - idx] = tmp;
        
        if (idx === Math.floor(s.length / 2)) return;

        return this.reverseStringRecursion(s, ++idx, s[idx]);
    }
}

module.exports = ReverseStringExercises;

File: CSharp/Studies/Studies/Strings/ReverseStringExercises.cs
using System;

namespace Studies.Strings
{
    public static class ReverseStringExercises
    {
        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/127/strings/879/
        /// reverses the array of characters with O(1) space
        /// </summary>
        /// <param name="str"></param>
        public static void ReverseString(char[] str)
        {
            var len = str.Length - 1;

            for (var i = 0; i < Math.Floor((double)str.Length / 2); i++)
            {
                var tmp = str[i];
                str[i] = str[len - i];
                str[len - i] = tmp;
            }
        }

        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/127/strings/879/
        /// reverses the array of characters with O(1) space (recursive version)
        /// </summary>
        /// <param name="str"></param>
        public static void ReverseString2(char[] str)
        {
            ReverseStringRecursion(str, 0, str[0]);
        }

        private static void ReverseStringRecursion(char[] str, int idx, char tmp)
        {
            str[idx] = str[str.Length - 1 - idx];
            str[str.Length - 1 - idx] = tmp;

            if (idx == Math.Floor((double)str.Length / 2)) return;

            ReverseStringRecursion(str, ++idx, str[idx]);
        }
    }
}
